core/src/main/java/org/apache/calcite/plan/RelOptRules.java

```diff
@@ -157,18 +157,18 @@ private RelOptRules() {
       UnionMergeRule.INSTANCE,
       UnionMergeRule.INTERSECT_INSTANCE,
       UnionMergeRule.MINUS_INSTANCE,
-      ProjectToWindowRule.PROJECT,
+      ProjectToWindowRule.ProjectToLogicalProjectAndWindowRule.INSTANCE,
       FilterMergeRule.INSTANCE,
       DateRangeRules.FILTER_INSTANCE,
       IntersectToDistinctRule.INSTANCE);
 
   static final List<RelOptRule> ABSTRACT_RELATIONAL_RULES = ImmutableList.of(
-      FilterJoinRule.FILTER_ON_JOIN,
-      FilterJoinRule.JOIN,
+      FilterJoinRule.FILTER_ON_JOIN.get(),
+      FilterJoinRule.JOIN.get(),
```

Comment:
I am not sure that using suppliers is the most elegant way to get rid of deadlocks on class loading. As I understand the deadlock mechanics right, It is something extremely rare. What if we just deprecate using static fields 
```
FilterJoinRule.FILTER_ON_JOIN,
FilterJoinRule.JOIN
```
and force using static constants declared within the actual class (not a super class, i.e. FilterJoinRule) : 
```
FilterIntoJoinRule.INSTANCE,
JoinConditionPushRule.INSTANCE
```
What do you think?

Reply:
I have made the change you suggested. We'll see whether anyone runs into deadlocks.

Fixed in 497fa05.